### PR TITLE
[Not to be merged] Update to compile and run test_torch with 3.13t

### PR DESCRIPTION
**Pull request by albanD**
This PR includes:
- [x] https://github.com/pytorch/pytorch/pull/129827 Upgrade pybind to v2.13.1 (the first with NoGil fixes)
- [x] https://github.com/pytorch/pytorch/pull/130895 Upgrade sympy to be importable in 3.13
- [ ] Fix some number tests to properly handle list vs tuple (also tracked in https://github.com/pytorch/pytorch/issues/132450)
- [x] Fix Tensor and Storage resurrection code to use _Py_NewReference instead of Py_INCREF to ensure the full refcount state is correct and the object is owned by the current thread (so that it gets freed instantly when the refcount reaches 0)
- [x] Remove the Py_TRACE_REFS logic from the Tensor subclass_dealloc similarly to how it was done in the Storage version as it is broken since 3.9
- [x] Use new thread safe version to access registered_instances in pybind
- [x] Fix serialization code now that Pickler.persistent_id is read_only
- [x] https://github.com/pytorch/pytorch/pull/130341
- (no need to land this) Uses numpy-2.1.0.dev0
- [x] https://github.com/pytorch/pytorch/pull/137396

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by guangyey** [comment]
@albanD I see you would to upgrade sympy to 1.13 or later. Different sympy versions have different behavior.
FYI https://github.com/pytorch/pytorch/pull/130729

**Comment by albanD** [comment]
I will most likely split this and the pybind upgrade to separate PRs for sure.
Is the CI being all green a good signal that this upgrade is most likely ok? I'll open a separate PR with the relevant people to make sure.

**Review by Skylion007 — APPROVED** [review]

File: requirements.txt [hunk 1]
```diff
@@ -9,7 +9,7 @@ requests
 setuptools
 types-dataclasses
 typing-extensions>=4.8.0
-sympy
+sympy >= 1.13.0
```
Comment (XuehaiPan) [inline requirements.txt:12]: You will need to pin the version in `setup.py` and CI constraint files. See https://github.com/pytorch/pytorch/pull/130836/files for reference.

Comment (albanD) [reply requirements.txt:12]: Yes I was not planning on merging this one as-is and this is more to use as an end-to-end check that CI works.
I will rebase above both the pybind and sympy upgrades once they're both landed!

Thanks a lot for your help with all of this!

File: functorch/csrc/dim/dim.cpp [hunk 2]
```diff
@@ -751,6 +751,8 @@ struct Tensor : public mpy::base<Tensor> {
     static mpy::obj<Tensor> create() {
         if (!TensorType) {
             TensorType = (PyTypeObject*) mpy::import("functorch.dim").attr("Tensor").ptr();
+            // NB: leak
+            Py_INCREF(TensorType);
```
Comment (albanD) [inline functorch/csrc/dim/dim.cpp:755]: We should use `.release()` above instead of this.

File: torch/serialization.py [hunk 3]
```diff
@@ -999,8 +1003,12 @@ def persistent_id(obj):
 
     # Write the pickle data for `obj`
     data_buf = io.BytesIO()
-    pickler = pickle_module.Pickler(data_buf, protocol=pickle_protocol)
-    pickler.persistent_id = persistent_id
+
+    class PyTorchPickler(pickle_module.Pickler):
```
Comment (albanD) [inline torch/serialization.py:1007]: I'm confused at the lint failure on this one haha
The one above is ok

**Comment by XuehaiPan** [comment]
FYI, there is a tracker for optree Python 3.13t support (3.13 is already supported via sdist).

- https://github.com/metaopt/optree/pull/137#issuecomment-2323242668

**Comment by albanD** [comment]
Ho no need to update this one, the key pieces were landed so closing this PR!
Will revive the final numpy 2.0 changes in the context of 3.13 ci enablement (not 3.13t).

**Pull request closed without merging**